File: HomeWork1/Hw1/Parser.cs
namespace Hw1;

public static class Parser
{
    public static void ParseCalcArguments(string[] args,
        out double val1,
        out CalculatorOperation operation,
        out double val2)
    {
        if (!IsArgLengthSupported(args)) throw new ArgumentException();
        try
        {
            val1 = double.Parse(args[0]);
            val2 = double.Parse(args[2]);
        }

        catch
        {
            throw new ArgumentException();
        }

        operation = ParseOperation(args[1]);
        if (operation == CalculatorOperation.Undefined)
            throw new ArgumentOutOfRangeException();


    }

    private static bool IsArgLengthSupported(string[] args) => args.Length == 3;

    private static CalculatorOperation ParseOperation(string arg)
    {
        switch (arg)
        {
            case "+":
                return CalculatorOperation.Plus;
            case "-":
                return CalculatorOperation.Minus; ;
            case "*":
                return CalculatorOperation.Multiply;
            case "/":
                return CalculatorOperation.Divide;
            case "_":
                return CalculatorOperation.Undefined;
            default:
                throw new InvalidOperationException();
        }
    }
}

File: Homework4/Hw4/Parser.fs
module Hw4.Parser

open System
open Hw4.Calculator


type CalcOptions = {
    arg1: float
    arg2: float
    operation: CalculatorOperation
}

let isArgLengthSupported (args : string[]) = args.Length = 3

let parseOperation (arg : string) =
    match arg with
    | "+" -> CalculatorOperation.Plus
    | "-" -> CalculatorOperation.Minus
    | "*" -> CalculatorOperation.Multiply
    | "/" -> CalculatorOperation.Divide
    | _ -> raise (ArgumentException())
    
let parseCalcArguments(args : string[]) =
    if isArgLengthSupported args = false
    then
        raise (ArgumentException())
    let parse1, value1 = Double.TryParse(args[0])
    let parse2, value2 = Double.TryParse(args[2])
    
    if  (parse1 && parse2) = false
    then
        raise (ArgumentException())
        
    let value3 = parseOperation(args[1])
    {
        arg1 = value1
        arg2 = value2
        operation = value3
    }